File: libs/ui_composite/src/tool_page_dialog.cpp
#include <pwb/ui_composite/tool_page_dialog.hpp>

#include <QCloseEvent>
#include <QDialogButtonBox>

namespace pwb::ui_composite {

ToolPageDialog::ToolPageDialog(QWidget* parent) : QDialog(parent) {
    setObjectName("ToolPageDialog");
    setModal(false);
    setWindowFlag(Qt::Window, true);
    resize(1080, 720);
    setMinimumSize(720, 480);
    auto* layout = new QVBoxLayout(this);
    layout->setContentsMargins(0, 0, 0, 8);
    body_ = new QVBoxLayout();
    body_->setContentsMargins(0, 0, 0, 0);
    layout->addLayout(body_, 1);
    auto* buttons = new QDialogButtonBox(QDialogButtonBox::Close, this);
    connect(buttons, &QDialogButtonBox::rejected, this,
            &ToolPageDialog::reject);
    layout->addWidget(buttons);
}

void ToolPageDialog::present(QWidget* page, const QString& title,
                             QWidget* home) {
    if (page == page_ && isVisible()) {
        setWindowTitle(title);
        raise();
        activateWindow();
        return;
    }
    release_page();
    page_ = page;
    home_ = home;
    body_->addWidget(page);
    page->show();
    setWindowTitle(title);
    show();
    raise();
    activateWindow();
}

void ToolPageDialog::release_page() {
    QWidget* page = page_;
    QWidget* home = home_;
    page_ = nullptr;
    home_ = nullptr;
    if (page == nullptr) return;
    body_->removeWidget(page);
    if (home != nullptr && home->layout() != nullptr)
        home->layout()->addWidget(page);
    else if (home != nullptr)
        page->setParent(home);
}

void ToolPageDialog::closeEvent(QCloseEvent* event) {
    release_page();
    QDialog::closeEvent(event);
}

void ToolPageDialog::reject() {
    release_page();
    QDialog::reject();
}

}  // namespace pwb::ui_composite

File: legacy/python_reference/product/paleo_workbench/ui/workstation/tool_page_dialog.py
"""Modeless dialog host for tool pages (数据制备 / 成图审核).

These pages are workflows over the central 编图 document. They must not
occupy a dock that competes with the map; a closable dialog keeps the
canvas in place.
"""
from __future__ import annotations

from PySide6.QtCore import Qt
from PySide6.QtWidgets import (
    QDialog,
    QDialogButtonBox,
    QVBoxLayout,
    QWidget,
)


class ToolPageDialog(QDialog):
    """Hosts one tool page, restoring it to its hub stack on close."""

    def __init__(self, parent: QWidget | None = None) -> None:
        super().__init__(parent)
        self.setObjectName("ToolPageDialog")
        self.setModal(False)
        self.setWindowFlag(Qt.WindowType.Window, True)
        self.resize(1080, 720)
        self.setMinimumSize(720, 480)
        layout = QVBoxLayout(self)
        layout.setContentsMargins(0, 0, 0, 8)
        self._body = QVBoxLayout()
        self._body.setContentsMargins(0, 0, 0, 0)
        layout.addLayout(self._body, 1)
        buttons = QDialogButtonBox(QDialogButtonBox.StandardButton.Close, self)
        buttons.rejected.connect(self.reject)
        layout.addWidget(buttons)
        self._page: QWidget | None = None
        self._home: QWidget | None = None

    def present(self, page: QWidget, title: str, home: QWidget | None) -> None:
        if page is self._page and self.isVisible():
            self.setWindowTitle(title)
            self.raise_()
            self.activateWindow()
            return
        self._release()
        self._page = page
        self._home = home
        self._body.addWidget(page)
        page.show()
        self.setWindowTitle(title)
        self.show()
        self.raise_()
        self.activateWindow()

    def _release(self) -> None:
        page = self._page
        home = self._home
        self._page = None
        self._home = None
        if page is None:
            return
        self._body.removeWidget(page)
        if home is not None and hasattr(home, "addWidget"):
            home.addWidget(page)
        elif home is not None:
            page.setParent(home)

    def closeEvent(self, event) -> None:  # noqa: N802
        self._release()
        super().closeEvent(event)

    def reject(self) -> None:
        self._release()
        super().reject()
